File: backend/routers/usuarios.py
"""
Router de usuarios del sistema. Listado (para selectores) + alta con rol asignable.
La autenticación vive en routers/auth.py; acá el alta va gateada con USUARIOS + WRITE.
"""
from uuid import UUID

from fastapi import APIRouter, Depends, Request

from integrations.supabase_client import supabase_admin
from schemas.usuario import (
    CambiarPasswordRequest,
    CambiarPasswordResponse,
    CrearUsuarioRequest,
    CrearUsuarioResponse,
)
from services.usuario_service import UsuarioService
from utils.permisos import Accion, Seccion, require_permission

router = APIRouter()
SECCION = Seccion.USUARIOS


def _svc() -> UsuarioService:
    return UsuarioService()


@router.get("", dependencies=[Depends(require_permission(SECCION, Accion.READ))])
async def list_usuarios(request: Request) -> dict:
    """Retorna usuarios activos del sistema (para el selector de responsable de objetivos)."""
    data = (
        supabase_admin.table("users")
        .select("id, nombre, apellido, email, username, rol")
        .eq("activo", True)
        .order("apellido")
        .execute().data or []
    )
    return {"items": data, "total": len(data)}


@router.post("", response_model=CrearUsuarioResponse, status_code=201,
             dependencies=[Depends(require_permission(SECCION, Accion.WRITE))])
async def crear_usuario(
    request: Request,
    body: CrearUsuarioRequest,
    service: UsuarioService = Depends(_svc),
) -> CrearUsuarioResponse:
    """Crea un usuario con el rol indicado (validado en el schema) y contraseña temporal. Solo admin_rrhh."""
    creado_por = request.state.user.get("id", "system")
    return service.crear_usuario(body, creado_por)


@router.post("/cambiar-password", response_model=CambiarPasswordResponse)
async def cambiar_password(
    request: Request,
    body: CambiarPasswordRequest,
    service: UsuarioService = Depends(_svc),
) -> CambiarPasswordResponse:
    """Cambia la contraseña del usuario autenticado (self-service, SIN gate de rol:
    cualquier usuario cambia SU propia clave). El id sale del token, nunca del body."""
    service.cambiar_password(request.state.user["id"], body.password_actual, body.password_nueva)
    return CambiarPasswordResponse()


@router.delete("/{user_id}", status_code=204,
               dependencies=[Depends(require_permission(SECCION, Accion.WRITE))])
async def eliminar_usuario(
    user_id: UUID,
    request: Request,
    service: UsuarioService = Depends(_svc),
) -> None:
    """Elimina un usuario del sistema. Solo admin_rrhh. No permite auto-eliminación.
    El id sale del path; el ejecutor, del token."""
    service.eliminar_usuario(str(user_id), request.state.user.get("id"))

File: frontend/services/usuarios.ts
import { apiFetch } from "@/services/api"
import type { EmpleadoListResponse } from "@/types/empleado"

/** Usuario del sistema, para selectores (espeja GET /api/usuarios del backend). */
export interface UsuarioOption {
  id: string
  nombre: string
  apellido: string
  email: string
  username: string
  rol: string
}

export interface UsuarioListResponse {
  items: UsuarioOption[]
  total: number
}

/** Payload de alta de un usuario mandos_medios. El rol lo fuerza el backend. */
export interface CrearUsuarioPayload {
  nombre: string
  apellido: string
  email: string
  username: string
  empleado_id?: string // opcional: vincula el user a su registro de empleado (líder)
}

/** Respuesta del alta: la contraseña temporal se muestra UNA sola vez. */
export interface CrearUsuarioResult {
  id: string
  username: string
  password_temporal: string
}

/** Empleado líder para el selector de vinculación (subconjunto de Empleado). */
export interface EmpleadoLider {
  id: string
  nombre: string
  apellido: string
  legajo: string | null
}

/** Lista los usuarios activos del sistema (para el selector de operador en filtros). */
export async function fetchUsuarios(): Promise<UsuarioListResponse> {
  return apiFetch<UsuarioListResponse>("/api/usuarios")
}

/** Crea un usuario mandos_medios; devuelve la contraseña temporal (no recuperable). */
export async function crearUsuario(payload: CrearUsuarioPayload): Promise<CrearUsuarioResult> {
  return apiFetch<CrearUsuarioResult>("/api/usuarios", {
    method: "POST",
    body: JSON.stringify(payload),
  })
}

/** Elimina un usuario del sistema por id (el backend bloquea la auto-eliminación). */
export async function eliminarUsuario(id: string): Promise<void> {
  await apiFetch<void>(`/api/usuarios/${id}`, { method: "DELETE" })
}

/** Cambia la contraseña del usuario autenticado (self-service). */
export async function cambiarPassword(passwordActual: string, passwordNueva: string): Promise<void> {
  await apiFetch<{ ok: boolean }>("/api/usuarios/cambiar-password", {
    method: "POST",
    body: JSON.stringify({ password_actual: passwordActual, password_nueva: passwordNueva }),
  })
}

/**
 * Empleados líderes activos para el selector de vinculación, de la empresa activa
 * (o todas si el selector global está en "todas"). Reusa GET /api/empleados con es_lider=true.
 */
export async function fetchEmpleadosLideres(): Promise<EmpleadoLider[]> {
  const res = await apiFetch<EmpleadoListResponse>(
    "/api/empleados?es_lider=true&estado=activo&page_size=100",
  )
  return res.items.map((e) => ({ id: e.id, nombre: e.nombre, apellido: e.apellido, legajo: e.legajo }))
}
